### DISCUSSION TITLE
Building highlights development discussion

### DISCUSSION THREAD

**Discussion by Zakru** [DigitalExtinction/Game#480] [General]
I've been prototyping the second part of #36, highlighting buildings with a square.

![screenshot of building highlights](https://github.com/DigitalExtinction/Game/assets/23387227/c2bea368-e3c1-4b03-bce1-7520ef269eaf)

So far I have a working prototype in [Zakru:feature/36](https://github.com/Zakru/DigitalExtinction-Game/tree/feature/36), screenshot above, although the shader part has no optimization data structure like the circle markers do, and the size should be made to fit around the building more snugly (it uses circular radius as ½ side length currently). I wanted to open this discussion to talk about reuse of code that already exists for the circle markers.

In terms of engine code, I think what I currently have in the branch is pretty decent. I extracted the `VisibilityFlags` from `CircleMarker` into an independent `MarkerVisibility` component, and made all of the code that only handles the visibility (most marker-related systems) use only that. Then I [created a `Marker` trait](https://github.com/Zakru/DigitalExtinction-Game/blob/8e8ba6847e1a55d7ee7ffed94979a65214662aec/crates/terrain/src/marker.rs#L41-L47) so that the `update_markers` system which updates relevant markers to `TerrainMaterial`s can easily be generic over marker types.

The shader and intermediate part are what I'd like to talk about. [Here is the diff of the current implementation.](https://github.com/Zakru/DigitalExtinction-Game/compare/main...feature/36#diff-83c09b89d2265d602d3e78a05f79ef79ea0d3bf1959db5dfe7bae57ae63c0218) I opted to quickly just use an array for prototyping before deciding how to reuse the data structures that already exist for circle markers, so there is no optimization at play. I haven't worked with k-d trees before, but optimization data structures like that are something I am personally very interested in learning, so lmk if and how I should approach that.

While writing I figured out I can use the AABB from `Ichnography` to get a more rectangular estimate, although now it looks a little too tight IMO. Opinions?

![screenshot of AABB-based highlight](https://github.com/DigitalExtinction/Game/assets/23387227/276b40bc-3c68-42f2-9fdd-f3e542fa0c9c)

Also please tell me if you'd rather discuss this over a draft PR or something, I just figured this might end up being a lengthy post. Or if I'm being silly and this could just be improved later 😄

**Comment by Indy2222** [comment]
Beware of #216. Current implementation, even with the kd-tree, is not very performant. We might need to take a different approach. An idea is to render the outlines outside of the terrain shader and somehow utilize instancing. This does not imply that your implementation cannot proceed, just make sure that even when many buildings are selected, it still performs reasonably well (i.e. 60+ FPS on a new mid-range GPU).

**Reply by Zakru** [reply]
Yes, I noticed this. I'll make sure, will just have to wait for build because I'd only build the default debug profile before, which... isn't that smooth, let's just say :D

**Reply by JackCrumpLeys** [reply]
You should use the testing profile - it has decent optimization to run at 60+ fps (at least on my laptop) but does not a take centuries to compile. (`cargo run ... --profile testing` ). You can have a read of the flags in the main `cargo.toml` (under `[profile.testing]`).

**Comment by Indy2222** [comment]
> While writing I figured out I can use the AABB from Ichnography to get a more rectangular estimate, although now it looks a little too tight IMO. Opinions?

I would use 2D AABB from `Ichnography` and expand it by a bit (a constant).

**Comment by Indy2222** [comment]
> Also please tell me if you'd rather discuss this over a draft PR or something, I just figured this might end up being a lengthy post. Or if I'm being silly and this could just be improved later smile

I think that having a discussion in a PR over concrete code would be easier.

**Reply by Zakru** [reply]
I shall open one then.

**Reply by Indy2222** [reply]
For reference #482.
